Sources/SPMBuildCore/ExtensionEvaluator.swift

```diff
@@ -0,0 +1,328 @@
+                    let displayName = cmd.displayName
+                    let execPath = AbsolutePath(cmd.executable)
+                    let arguments = cmd.arguments
+                    let workingDir = cmd.workingDirectory.map{ AbsolutePath($0) }
+                    let environment = cmd.environment
+                    switch extTarget.capability {
+                    case .prebuild:
+                        let derivedSourceDirPaths = cmd.derivedSourcePaths.map{ AbsolutePath($0) }
+                        return .prebuildCommand(displayName: displayName, execPath: execPath, arguments: arguments, workingDir: workingDir, environment: environment, derivedSourceDirPaths: derivedSourceDirPaths)
+                    case .buildTool:
+                        let inputPaths = cmd.inputPaths.map{ AbsolutePath($0) }
+                        let outputPaths = cmd.outputPaths.map{ AbsolutePath($0) }
+                        let derivedSourcePaths = cmd.derivedSourcePaths.map{ AbsolutePath($0) }
+                        return .buildToolCommand(displayName: displayName, execPath: execPath, arguments: arguments, workingDir: workingDir, environment: environment, inputPaths: inputPaths, outputPaths: outputPaths, derivedSourcePaths: derivedSourcePaths)
+                    case .postbuild:
+                        return .postbuildCommand(displayName: displayName, execPath: execPath, arguments: arguments, workingDir: workingDir, environment: environment)
+                    }
+                }
+                
+                // Create an evaluation result from the usage of the extension by the target.
+                let textOutput = String(decoding: emittedText, as: UTF8.self)
+                evalResults.append(ExtensionEvaluationResult(extension: extTarget, commands: commands, diagnostics: diagnostics, textOutput: textOutput))
+            }
+            
+            // Associate the list of results with the target.  The list will have one entry for each extension used by the target.
+            evalResultsByTarget[target] = evalResults
+        }
+        return evalResultsByTarget
+    }
+    
+    
+    /// Private helper function that serializes an ExtensionEvaluationInput as input JSON, calls the extension runner to invoke the extension, and finally deserializes the output JSON it emits to a ExtensionEvaluationOutput.  Adds any errors or warnings to `diagnostics`, and throws an error if there was a failure.
+    /// FIXME: This should be asynchronous, taking a queue and a completion closure.
+    fileprivate func runExtension(sources: Sources, input: ExtensionEvaluationInput, extensionRunner: ExtensionRunner, diagnostics: DiagnosticsEngine, fileSystem: FileSystem) throws -> (output: ExtensionEvaluationOutput, stdoutText: Data) {
+        // Serialize the ExtensionEvaluationInput to JSON.
+        let encoder = JSONEncoder()
+        encoder.outputFormatting = [.prettyPrinted, .withoutEscapingSlashes]
```

Comment:
iirc we have the helpers for this in Basics

Reply:
Right, that's a good point.  Also, this doesn't need to be pretty-printed, so I should probably just take this out.